### DISCUSSION TITLE
How to limit array index and case number to be int terminal in grammar?

### DISCUSSION THREAD

**Discussion by durianwaffle** [eclipse-langium/langium#1704] [Q&A] [answered]
Hi Langium community,

I found that when I change the terminal NUMBER to INT_NUMBER. The grammar will show errors.
Do you know how to fix it in grammar?

[Langium playground for grammar of using int number as index in array that shows error](https://langium.org/playground?grammar=OYJwhgthYgBAgiAlgFwBYQKYqQYwM4BQhmAdiiAJ6wCyA9gCYCuANpgFyGzewAU%2BKMCkxZy%2BANQBeAMqDholAEoAVAG5isoSLIpOPWABFMAMySlUSOqVgAfWAFEAbmBZMhl0usJHT5nFb0eADEmUlx-UiNcFhh3K1suHlkQJnCAFUoAB0womPAIhP0ANRgkMAAjNlzYgq8QsIjq-I9A7gByY1DwjzbYUkhMSQBJAz423l6YYAlJAAVYrGE4CYAaSZBpqXnwRcwQFVg2xSOAfkT29l7MAA9MkEl7W5BMfHwPQ9U2r0Jk1JQM7JNOLWZ4oJggUj4WAAOQGDHsbAUrVgAG8zMY9rBfuksjlMNEah4AL7nQ4oXEMXr9LDDUZtS6wUltAQpcJtUlYCDlPYzLEUP40ETckDKJlkBgAfRZf16X163EIXhKyAqVXxeWBrWkSAgmTYyrKlTxBOa8TsiHAlANquNGoKNmSZmA1qNQIiio0Or1mBdapNwNgoPBkJhcIR2nIyLRpAxcF9tsJViJjP01MGIyZDPJ2UkAMwACUTHsyLhMCmeLw0GB8EM-GUWCVXJgTpJ6a3YM4mw8ni83lZFCdy9xK9X82BSMBm5JDhMsUxyuAJ2WjqcPYQLWAraUbW73kGIVDYVh4YidFH0Zj47uk0O%2BgNaSmAPSPnj03q8bODPOF2MlzCKYc2gAbV6BgdTIPtSCkaR50XScxjWMDREg6DYPHScDjaABdVd9DfQ4YEtWUQNgJCII8VCF3QstVjaMjIQomQ0KXTCcMOOhjF6T9c1xH9izCMsFX0EcazrFxGyYKc216TtJOnDdKFrCwXCQAAvYEB30NcFKUnAVPUiJkUOEjMFPMQpEeO5e3eWjYFMiMUBmSznled5WPZLxHQnK91UTEFsGDQ8wzM3RSWjWNYB8-0ImTUk0wfTN324rzgAlPNYAA4DejYCd0EkaEAFUaAAIXsfNsPZYSq1E5SGxcSSW2kjt6sGPlkAnZzrP7M5PIoJ1OtcgJmqbWBp2kCVisoYQJQAYSrEAJWSGwDAmqbMFm%2BbFooLwlggMwXCxVbprmmAtpATgAEg2gAHTaXpxsm47NokzBlEOW6vmIXb9pYQwjvWk6FuSS62gAIl6FbHoB56WresHPsIT8EHKFkwHCHwzAseJpwxuszVgbYBiWLwnHqzVSRuKzBtIbsqcgj4EYGyDkXgBgwIiLwCv6KgmZac46GyfI6HuCYAEJejsNoAFoJcOcQjmGuTeasD1oSYLk9mV0h2EVwZCpKsr1GKug6DYcctZ12SpP5TBZY6Fx8FtrxBQ1kAZpcX6LxAKELdJWZkGgKhCAu3hwsxF3hXdlgWAAOisxxLCYfBJFwcFnnIZM2hjq4QskICj0wE8HPYEYsNJFQPSRgui4UUaCYWbBMTsepuisa9rDsKK7XeOxsRQCOm7av483b2BetZfuhT2ZF4pGQ4s1xHjsj49PSwZhILtn0ZEr4ZK%2BonNLcQy4zsrIYA8v10ryrYuU12%2B-pfuMcBgFrgAJG4JQMJBgFQdhHyAgADFLAAnGAKWxh4BSyCFhR8O09h7QfrAJ%2BYAX46EOlDDap0XqXRmnQKAVhMELRerYQ4AASD6JD4aUNIb0d%2B1xP7f1QLAOhDCf4oDgSABBB1kGoPIH9DBgMJTYODrg-BpBCFCJapQihktSHg2obQj%2BX82HMKUYwlAqj6HKKYSw7R7CvrwJ%2Bkg5%2BtdREQAIYI4RV1DiUPFpQgAxHbAApHbAAZLLf%2B11eBS0fDAkh-8gHAL8XYf%2B7ApYAAFgmwH-pAgAWlEzxQEpYAAMElATAapNJ10URSwAH4wODrImh1CAAydtSHQnKbMcp5VqFpARjBKiS5YBex9tcVAmh5BnkSGHOAjS4KYBJNwJ%2BeDhjQjSBKS%2BZVDgx2zrAFAdAxkTKmfmNcKVD7ZEti1SQEwWROjaDYNoAB3PZE4jheEJrsc6d4aRz3wtxb8RZV6YC8I838Akdb6AuqSN4wBxFmBQBKbiEwAUHOZKCw5YFyBgpYKCxQpIbCklCL8-55AgWLwmEnCFbQmDYqYFClAYLWBwoRaSZ4Lh0U5gmOSlgMKaVHFJRdBgWgcBYEpa1CYrLbaQq0GCrlEoOISmZZQMFzLprjklFyhl%2BhnyIp4M%2BNqToNlllJfocoJsWDsp2W0dVptpU8DlTwcoa18BaomMa4QYKjnC0pJC61IBbVtBYPayk8KLqIouuOSgZq2herBV6iUhFNwMvdWSp5f5pz52CsXUujIvDFX2jzHs1MtldgTdzb1tZ9HrjZljawrSHDJuZqSGgrAcB6jwAGUOXtYDppgJQLWMc2DGBQCnNOOhkwCz2EIYW2r5aUJlgBZAwA0CttLSwctsLcAaTUMQcdk7K0FALb7fQlkrA6DKAUatMZMR1qTXTDwTaTCttTiAdOKBO2Cx7SLNoyg7aPgVsO0dtN13kE3R4FQJNbivr0kund3tC0HoCCWxgrA6B8F6bWxNDai2HubSe9tGdYBdqFvcNoAA9XoT6x2gZYHQT9c7cPgeXbB4D%2Bh-Y6nrVrCDNa90waA6QI9La21no7chq9Cy0MuMDN-Z9sw6BHII4QfjRyWn-raQx5EA83Yexo-%2BqDGbG3wZY%2Bey9khQbKGUKDHjI6x1Txk9HITFHA70ZcvTEjkmmSzmoyuQoPBINq1diU1A3aWDJitvlIqV87PcEgy3CIUc3NIK6HneAKMKBoxQLjPNWExizimE5Uj1hbIJYsklzCxwBwXIDvWsTsYJNmb5nhazSXDjHB87ALmVHSt2B4KSY2ptMDmxq6SRzwpqOGu4HRrWXB1D3wOnRiUWaQYACZHFjAlG0QcCalDiD64YxBPDa4zZ1gEqWABGGBqhvloCQGzMg8yFsHQAOrSFW9dCQsDEZHd%2BiMVbQFA1S1UnErCQFrpHIlFhZQV3%2Bu-RWYGAKB4%2Bjq2FPdwJWFxC8GujHQBICvsDh%2Bzd2AQxxmTK89M-cIZSAg%2BntE2HQTxBXd2-t6wv3aAlNmgAeRoDQew4zzuPmusoN7%2BBrrSC%2BycJn11nzqGJwwA7ZPpAU5mtT2n9Ponc%2B50BdD11SDXRAF92BQA&content=IYLgBMBOnAng2gRgHTIEwF0wHsBmYBLAOwBcAoMgYxGJLBAF5gwBqMAIwvZHe2wBt6DMAA9WYWBVwBXIpRIFsRMLgAUoWgBpuvAQEoQlANxA)
![image](https://github.com/user-attachments/assets/40956014-94d4-491e-ba2b-0c9fa46ac2be)

[Langium playground for grammar of using float number as index in array that doesn't show error](https://langium.org/playground?grammar=OYJwhgthYgBAgiAlgFwBYQKYqQYwM4BQhmAdiiAJ6wCyA9gCYCuANpgFyGzewAU%2BKMCkxZy%2BANQBeAMqDholAEoAVAG5isoSLIpOPWABFMAMySlUSOqVgAfWAFEAbmBZMhl0usJHT5nFb0eADEmUlx-UiNcFhh3K1suHlkQJnCAFUoAB0womPAIhP0ANRgkMAAjNlzYgq8QsIjq-I9A7gByY1DwjzbYUkhMSQBJAz423l6YYAlJAAVYrGE4CYAaSZBpqXnwRcwQFVg2xSOAfkT29l7MAA9MkEl7W5BMfHwPQ9U2r0Jk1JQM7JNOLWZ4oJggUj4WAAOQGDHsbAUrVgAG8zMY9rBfuksjlMNEah4AL7nQ4oXEMXr9LDDUZtS6wUltAQpcJtUlYCDlPYzLEUP40ETckDKJlkBgAfRZf16X163EIXhKyAqVXxeWBrWkSAgmTYyrKlTxBOa8TsiHAlANquNGoKNmSZmA1qNQIiio0Or1mBdapNwNgoPBkJhcIR2nIyLRpAxcF9tsJViJjP01MGIyZDPJ2UkAMwACUTHsyLhMCmeLw0GB8EM-GUWCVXJgTpJ6a3YM4mw8ni83lZFCdy9xK9X82BSMBm5JDhMsUxyuAJ2WjqcPYQLWAraUbW73kGIVDYVh4YidFH0Zj47uk0O%2BgNaSmAPSPnj03q8bODPOF2MlzCKYc2gAbV6BgdTIPtSCkaR50XScxjWMDREg6DYPHScDjaABdVd9DfQ4YEtWUQNgJCII8VCF3QstVjaMjIQomQ0KXTCcMOOhjF6T9c1xH9izCMsFX0EcazrFxGyYKc216TtJOnDdKFrCwXCQAAvYEB30NcFKUnAVPUiJkUOEjMFPMQpEeO5e3eWjYFMiMUBmSznled5WPZLxHQnK91UTEFsGDQ8wzM3RSWjWNYB8-0ImTUk0wfTN324rzgAlPNYAA4DejYCd0EkaEAFUaAAIXsfNsPZYSq1E5SGxcSSW2kjt6sGPlkAnZzrP7M5PIoJ1OtcgJmqbWBp2kCVisoYQJQAYSrEAJWSGwDAmqbMFm%2BbFooLwlggMwXCxVbprmmAtpATgAEg2gAHTaXpxsm47NokzBlEOW6vmIXb9pYQwjvWk6FuSS62gAIl6FbHoB56WresHPsIT8EHKFkwHCHwzAseJpwxuszVgbYBiWLwnHqzVSRuKzBtIbsqcgj4EYGyDkXgBgwIiLwCv6KgmZac46GyfI6HuCYAEJejsNoAFoJcOcQjmGuTeasD1oSYLk9mV0h2EVwZCpKsr1GKug6DYcctZ12SpP5TBZY6Fx8FtrxBQ1kAZpcX6LxAKELdJWZkGgKhCAu3hwsxF3hXdlgWAAOisxxLCYfBJFwcFnnIZM2hjq4QskICj0wE8HPYEYsNJFQPSRgui4UUaCYWbBMTsepuisa9rDsKK7XeOxsRQCOm7av483b2BetZfuhT2ZF4pGQ4s1xHjsj49PSwZhILtn0ZEr4ZK%2BonNLcQy4zsrIYA8v10ryrYuU12%2B-pfuMcBgFrgAJG4JQMJBgFQdhHyAgADFLAAnGAKWxh4BSyCFhR8O09h7QfrAJ%2BYAX46EOlDDap0XqXRmnQKAVhMELRerYQ4AASD6JD4aUNIb0d%2B1xP7f1QLAOhDCf4oDgSABBB1kGoPIH9DBgMJTYODrg-BpBCFCJapQihktSHg2obQj%2BX82HMKUYwlAqj6HKKYSw7R7CvrwJ%2Bkg5%2BtdREQAIYI4RV1DiUPFpQgAxHbAApHbAAZLLf%2B11eBS0fDAkh-8gHAL8XYf%2B7ApYAAFgmwH-pAgAWlEzxQEpYAAMElATAapNJ10URSwAH4wODrImh1CAAydtSHQnKbMcp5VqFpARjBKiS5YBex9tcVAmh5BnkSGHOAjS4KYBJNwJ%2BeD8pFSvocGO2dYAoDoGMg2%2BY1wpUPtkS2LVJATBZE6NoNg2gAHctkTiOF4Qmuxzp3hpHPfC3FvxFlXpgLwtzfwCR1voC6pI3jAHEWYFAEpuITB%2BTs5kgLdlgXIEClggLFCkhsKSUInzvnkD%2BYvCYScQVtCYOipgYKUBAtYFCmFpJnguGRTmCYxKWAQopUcQlF0GBaBwFgUlrUJiMttqCrQQK2USg4hKellAgX0umuOSUbKaX6GfLCngz42pOhWWWQl%2BhygmxYMyjZbRlWm3FTwKVPByhrXwGqiY%2BrhBAr2cLSkoLzUgEtW0Fg1rKTQourCi645KBGraG6oFbqJSEU3DS51RK7l-mnPnYKxdS6Mi8MVfaPMezUzWV2GN3N3W1n0euNmWNrCtIcPG5mpIaCsBwHqPAAZQ5e1gMmmAlAtYxzYMYFAKc046GTALPYQhhbqvlpQmWAFkDADQI2wtLBi2QtwBpNQxBh2jtLQUHNvt9CWSsDoMoBRy0xkxFWuNdMPB1pMI21OIB04oFbYLDtIs2jKDto%2BBW-bB202XeQVdHgVAk1uI%2BvSc6N3e1zTugIBbGCsDoHwXplbY01rzbu%2BtB7m0Z1gG2oW9w2gAD1eh3qHYBlgdBX1Tsw8B%2BdkH-36H9jqatWsQMVq3RBv9pA90NqbUelt8Gz2zKQy4wM3972zDoHsnDhBuN7Jad%2BtpNHkQDzdh7Cj36wMptrdBhjx7T2SFBsoZQoMOMDqHVPCT0c%2BMkcDtRly9MCOiaZLOcjK5Cg8FA2rV2JTUDtpYMmK28yJm6u4KBluEQo5OaQV0PO8AUYUDRigXGWasJjFnFMJyhHrC2WixZWLmFjgDhOQHatQnYwiaM3zPC5nYuHGOFZ7gXMyMFbsDwUkxtTaYHNuV0ktnhTkfczJsrNGuDqHvgdKjEo00gwAEyOLGBKNog4Y1KHEJ1wxiCeG13GzrAJUsACMMDVDvLQEgNmZAZnTYOgAdWkAt66EhYGI1279EYC2gK%2BqlqpOJWEgLXT2RKLCyhTtdd%2BpfMqgYAoHj6OrYUV3AlYXELwa6MdAEgNewOU7MqPuwCGNCNIEovv5h%2B2CP7pAAfT2iZDoJ4hTsba29YeHNASmzQAPI0BoPYJHR3HzXWUI9-A11pCvZOIz66z51BE4YNt%2BH0hyczSpzTun0Sudc6Ash66pBrogFe7AoAA&content=IYLgBMBOnAng2gRgHQuQJmQZgLpgPYBmYAlgHYAuAUFQMYjkVggC8wYA1GAEY3cjd8%2BADbMWYAB6cwsGoQCuZWhRL4yYQgApQjADT9BIgJQhaAbiA)
![image](https://github.com/user-attachments/assets/f9ea1706-575b-433c-bdbb-de3a09ef753b)

![image](https://github.com/user-attachments/assets/40876fec-afcd-422e-bf63-ac33de42b369)

**Comment by ydaveluy** [comment]
I think that you probably have a decimal number rule that match `1.`. If you add a space like `array[1 .. 2] does it solve the issue ?

**Reply by durianwaffle** [reply]
![image](https://github.com/user-attachments/assets/3325aac3-a58d-427b-87d7-0303b6a2c389)
Yes, it will solve this issue

**Reply by durianwaffle** [reply]
But what if users don't want to type space in array type declaration?
I want to support this `array[1..2] `

**Comment by msujew** [comment]
Hey @durianwaffle,

The `INT_NUMBER` terminal is shadowed by the `NUMBER` terminal since it appears first in the grammar.

**Reply by durianwaffle** [reply]
I see! Thank you @msujew, do you know how to fix this issue? 

I have changed NUMBER terminal to disable `a := 1`. or `a := .1` but it still have issues.
`terminal NUMBER returns number: /[0-9]+(\.[0-9]+)?([eE][+-]?[0-9]+)?/;`

[Langium playground for updated grammar of using int number as index in array that still shows error](https://langium.org/playground?grammar=OYJwhgthYgBAgiAlgFwBYQKYqQYwM4BQhmAdiiAJ6wCyA9gCYCuANpgFyGzewAU%2BKMCkxZy%2BANQBeAMqDholAEoAVAG5isoSLIpOPWABFMAMySlUSOqVgAfWAFEAbmBZMhl0usJHT5nFb0eADEmUlx-UiNcFhh3K1suHlkQJnCAFUoAB0womPAIhP0ANRgkMAAjNlzYgq8QsIjq-I9A7gByY1DwjzbYUkhMSQBJAz423l6YYAlJAAVYrGE4CYAaSZBpqXnwRcwQFVg2xSOAfkT29l7MAA9MkEl7W5BMfHwPQ9U2r0Jk1JQM7JNOLWZ4oJggUj4WAAOQGDHsbAUrVgAG8zMY9rBfuksjlMNEah4AL7nQ4oXEMXr9LDDUZtS6wUltAQpcJtUlYCDlPYzLEUP40ETckDKJlkBgAfRZf16X163EIXhKyAqVXxeWBrWkSAgmTYyrKlTxBOa8TsiHAlANquNGoKNmSZmA1qNQIiio0Or1mBdapNwNgoPBkJhcIR2nIyLRpAxcF9tsJViJjP01MGIyZDPJ2UkAMwACUTHsyLhMCmeLw0GB8EM-GUWCVXJgTpJ6a3YM4mw8ni83lZFCdy9xK9X82BSMBm5JDhMsUxyuAJ2WjqcPYQLWAraUbW73kGIVDYVh4YidFH0Zj47uk0O%2BgNaSmAPSPnj03q8bODPOF2MlzCKYc2gAbV6BgdTIPtSCkaR50XScxjWMDREg6DYPHScDjaABdVd9DfQ4YEtWUQNgJCII8VCF3QstVjaMjIQomQ0KXTCcMOOhjF6T9c1xH9izCMsFX0EcazrFxGyYKc216TtJOnDdKFrCwXCQAAvYEB30NcFKUnAVPUiJkUOEjMFPMQpEeO5e3eWjYFMiMUBmSznled5WPZLxHQnK91UTEFsGDQ8wzM3RSWjWNYB8-0ImTUk0wfTN324rzgAlPNYAA4DejYCd0GGaE0glaEAFUaAAIXsfNsPZYSq1E5SGxcSSW2kjsmsGPlkAnZzrP7M5PIoJ0etcgI2qbWBp2kCUysoYQJQAYSrEAJWSGwDGm2bMAWpaVooLwlggMwXCxDa5sWmBdpATgAEg2gAHTaXoppms6dokzBlEOB6vmIA6jpYQxTq287luSG62gAIl6daXuBt72s%2ByGfsIT8EHKFkwHCHwzAseJp2xuszVgbYBiWLwnCazVSRuKyRtIbtacgj5keGyDkXgBgwIiLxiv6KhWZac46GyfI6HuCYAEJejsNoAFppcOcQjjGuSBasD1oSYLk9jV0h2BVwYSvKyr1DKug6DYcddf12SpP5TAFY6Fx8AdrxBW1kB5pcAGLxAKFrdJWZkGgKhCGu3hwsxd3hS9lgWAAOisxxLCYfBJFwcFnnIZM2njq4QskICj0wE8HPYEYsNJFQPVR4vS4UCbiYWbBMTsepuisa9rDsKK7XeOxsRQaPW86v48y72ABtZIehT2ZF4pGQ4s1xHjsj4rPS2ZhJroX0ZEr4ZLBonNLcQy4zsrIYA8qNiqqrYuU1z%2B-oAeMcBgAbgAJG4JQMJBgFQdgj4gIAAZZYAE4wCy2MPAWWQQsKPn2nsQ6z9YCvzAO-HQJ1YbbQuu9G6806BQCsDg5a71bCHAACTfXIUjGhFDehf2uD-P%2BqBYCMOYf-FAiCQDIOOmgjB5BAbYJBhKPBYcCFENICQ0R7UaHUJlhQqGdCGHf1-pwthqiWEoA0UwtRrD2F6K4b9JB-1UFvwbhIiAxCRFiNuocGhUsaEAGJHYAFJHYADIFZALurwWWj54HkKAaAsBgS7BAPYLLAAAmE2AQCYEAC1Yk%2BKArLAABskoCkDVKZLuiiWWAA-eBYcFH0LoQAGUdhQ6EVTZhVKqnQtIyMYJUSXLAX2-trioE0PIM8iRI5wBaXBTAJJuCv0IflQqN9KqHHjnnWAKA6CTKKqVW%2Ba4Uon2yDbdqkgJgsidG0GwbQADu%2ByJxHC8CTXYV07w0kXvhbi34iwb0wF4J5v4BL630NdUkbxgBSLMCgCU3EJiAsOcyMFRywLkHBSwMFihSQ2FJKEP5ALyDApXhMVOkK2hMBxUwaFKBwWsHhYi0kzwXAYpzBMClLBYW0qOGS66DAtA4CwFSjqEw2UOyhVocF3KJQcQlCyyg4KWVzXHJKbljL9DPiRTwZ8nUnSbLLGS-Q5RzYsA5bstoGqLYyp4PKng5RNr4G1RME1whwXHLFpSKFNqQB2raCwB1lIEXXSRddcclBzVtG9eC71EpCKbkZR68lzy-zTiLsFMuFdGReDKkdfmPY6bbK7ImvmPraxGPXJzXG1gOkOBTWzUkNBWA4D1HgAMEdfawAzTASgut45sGMCgdOmcdDJmFnsIQYsdVKxofLACyBgBoDbWWlgFa4W4A0moYgE6p1VoKIWgO%2BhLJWB0GUAoNaYyYnrcmxmHhm0mDbRnEAWcUBdpFr28WbRlCO0fMrEdY6GYbvIFujwKhya3DfXpZdu6-ZFsPQEUtjBWB0D4AMutSbG3FqPS209Hbs6wG7aLe4bQAB6vRn3jrAywOgX7514YgyuuDIH9BBx1A23WkHa37tg8B0gx7W3tvPZ2lD17FnofcYGP%2BL7Zh0GOYRwgAnjntIA50xjyJh6e29rRgD0HM1NoQ6xi9V7JAQ2UMoCGvHR3jtnrJuOwnKMhwYy5JmpGpNMlnDRlchQeBQc1h7cpqAe0sGTLbSQ0z8z2e4FB9uERY7udQV0Qu8B0YUExigAm%2BasJjFnFMJyZHrC2USxZZLmFjgDkucHBt4nYySfM4LPCNnkuHGOL52AvNqNlbsDwUkZsLaYCtrV0kTnhQ0aNdwejusuDqCfsdejEps3gwAEwuLGBKNog5E1KHEP1kxKD%2BEN1m-rYJssACM8DVA-LQEgTmZAFmLeOgAdWkGtu6EgEEo2OwDEYa2gJBtlqpRJWEgJ3WORKLCyhrsDYBt5wMAUDx9C1sKB7ISsLiF4HdeOIDwGQ4HL927sAhgFRWcbHz%2B4QykFB3POJcPQniGu3tg71g-u0HKQtAA8jQGg9gCoXcfHdZQ738B3WkN9k4zO7rPnUCThgh3yfSEp-NGndOGdxJ5zzoCGG7qkDuiAb7CCgA&content=IYLgBMBOnAng2gRgHTIEwF0wHsBmYBLAOwBcAoMgYxGJLBAF5gwBqMAIwvZHe2wBt6DMAA9WYWBVwBXIpRIFsRMLgAUoWgBpuvAQEoQlANxA)

**Reply by msujew** [reply, marked as answer]
You should just use a single number terminal. You can later perform validations in the appropriate places that the entered number is an integer (and not some `real` value).

**Reply by durianwaffle** [reply]
I see! Thank you @msujew. 
Since I can only use a single number terminal, can't limit the type of number in grammar. 
I will **do validation for that (to check it's not real)**
